packages/eslint-plugin/docs/rules/no-explicit-any.md

```diff
@@ -158,7 +158,24 @@ interface Garply {
 
 ## When Not To Use It
 
-If an unknown type or a library without typings is used
+**Higher Order Functions**. If you're typing a higher order function such as `compose`, `pipe`, etc, you may need `any` because TypeScript types are not fully co-expressive with idiomatic functional JavaScript. Such functions are intentionally generic, and you can assign explicit types later to the returned functions, e.g.:
+
+```TypeScript
+// eslint-disable-next-line no-explicit-any
+type a2a = (x: any) => any;
+type compose = (...fns: a2a[]) => a2a;
+const pipe: compose = (...fns) => x => fns.reduce((y, f) => f(y), x);
+const compose: compose = (...fns) => x => fns.reduceRight((y, f) => f(y), x);
+
+type n2n = (n: number) => number;
+const g: n2n = n => n + 1;
+const f: n2n = n => n * 2;
+
+const h: n2n = pipe(g, f);
+const j: n2n = compose(f, g);
```

Comment:
It's not hard to make this type-safe:

```ts
type a2a<T> = (x: T) => T;
const pipe = <T,>(...fns: a2a<T>[]) => (x: T) => fns.reduce((y, f) => f(y), x);
const compose = <T,>(...fns: a2a<T>[]) => (x: T) => fns.reduceRight((y, f) => f(y), x);

type n2n = (n: number) => number;
const g: n2n = n => n + 1;
const f: n2n = n => n * 2;

const h: n2n = pipe(g, f);
const j: n2n = compose(f, g);
```

So I don't think this should be shown. To date, I haven't seen a single case where you cannot express patterns fully type-safely (or, if so, the pattern is probably awkward in the first place anyway), if you think hard enough about it.

Reply:
Those are not correct types for compose and pipe, which can take any number of functions of different types:

```
type a2a<T> = (x: T) => T;
const pipe = <T,>(...fns: a2a<T>[]) => (x: T) => fns.reduce((y, f) => f(y), x);
const compose = <T,>(...fns: a2a<T>[]) => (x: T) => fns.reduceRight((y, f) => f(y), x);

type n2n = (n: number) => number;
const g: n2n = n => n + 1;
const f: n2n = n => n * 2;
type n2s = (n: number) => string;
const exclaim:n2s = (n) => `${n}!`;

const h: n2s = pipe(g, f, exclaim);
const j: n2s = compose(exclaim, f, g);
```

Errors:

```
Type '(x: number) => number' is not assignable to type 'n2s'.
  Type 'number' is not assignable to type 'string'.
Argument of type 'n2s' is not assignable to parameter of type 'a2a<number>'.
  Type 'string' is not assignable to type 'number'.
Type '(x: number) => number' is not assignable to type 'n2s'.
  Type 'number' is not assignable to type 'string'.
Argument of type 'n2s' is not assignable to parameter of type 'a2a<number>'.
```

vs

```
type a2a = (x: any) => any;
const pipe = (...fns: a2a[]) => (x: any) => fns.reduce((y, f) => f(y), x);
const compose = (...fns: a2a[]) => (x: any) => fns.reduceRight((y, f) => f(y), x);

type n2n = (n: number) => number;
const g: n2n = n => n + 1;
const f: n2n = n => n * 2;
type n2s = (n: number) => string;
const exclaim:n2s = (n) => `${n}!`;

const h: n2s = pipe(g, f, exclaim);
const j: n2s = compose(exclaim, f, g);
```

No errors.

Reply:
FWIW, most FP libraries define overloads for 1–10 piped functions and _then_ as a last resort have the `(...fns: (a: any) => any) => any` signature. This does seem awkward but you don't take type safety away from your consumers.

Reply:
1. It is *extremely common* to declare compose and pipe inline because they are convenient one-liners that many functional programmers can crank out faster than they can import a library. Inline definitions also reduce dependencies and are usually far simpler than the library alternatives (including typings).

Even if that could work, even doing it for all the common basic combos like `string => Array<Number>` quickly leads to an untenable combinatorial explosion. In other words, the "additional type safety" provided by overloading is a fragile illusion:

* string
* number
* bigint
* boolean
* symbol
* object
* Array<T>
* Promise<T>
* Map<K, V>
* WeakMap<K extends object, V>
* Set<T>
* WeakSet<T>
* Date
* RegExp
* Error
* Function (oh no! functions can take and return functions! Inception time...)
... -- Remember, we have to raise this whole list to an exponent, and getting into all the combinations of those <T> is just not a workable solution.

Reply:
I think you misunderstood about overloads. I'm talking about these:

```ts
function pipe<T1, R>(f1: (a: T1) => R): R;
function pipe<T1, T2, R>(f1: (a: T1) => T2, f2: (a: T2) => R): R;
function pipe<T1, T2, T3, R>(f1: (a: T1) => T2, f2: (a: T2) => T3, f3: (a: T3) => R): R;
```

Reply:
Ah, but those don't work because the return value of pipe is a function of a type from `A => C for A => B |> B => C`:

```
// eslint-disable no-explicit-any
// type a2a = (x: any) => any;
// const pipe = (...fns: a2a[]) => (x: any) => fns.reduce((y, f) => f(y), x);
function pipe<T1, R>(f1: (a: T1) => R): R;
function pipe<T1, T2, R>(f1: (a: T1) => T2, f2: (a: T2) => R): R;
function pipe<T1, T2, T3, R>(f1: (a: T1) => T2, f2: (a: T2) => T3, f3: (a: T3) => R): R;
function pipe(...fns: Array<(...args: any[]) => any>): (...args: any[]) => any {
  return (input: any) => fns.reduce((prev, fn) => fn(prev), input);
}
// eslint-enable no-explicit-any

type n2n = (n: number) => number;
const g: n2n = n => n + 1;
const f: n2n = n => n * 2;
type n2s = (n: number) => string;
const exclaim:n2s = (n) => `${n}!`;
```

Error:
```
Type 'string' is not assignable to type 'n2s'.
```

We can try to fix it by expecting the return value to be a function, but TypeScript can't automatically infer the correct function type to expect:

```
function pipe<T1, R>(f1: (a: T1) => R): (a: T1) => R;
function pipe<T1, T2, R>(f1: (a: T1) => T2, f2: (a: T2) => R): (a: T1) => R;
function pipe<T1, T2, T3, R>(f1: (a: T1) => T2, f2: (a: T2) => T3, f3: (a: T3) => R): (a: T1) => R;
function pipe(...fns: Array<(...args: any[]) => any>): (...args: any[]) => any {
  return (input: any) => fns.reduce((prev, fn) => fn(prev), input);
}
// eslint-enable no-explicit-any

type n2n = (n: number) => number;
const g: n2n = n => n + 1;
const f: n2n = n => n * 2;
type n2s = (n: number) => string;
const exclaim:n2s = (n) => `${n}!`;
type s2b = (s: string) => boolean;
const isHuge:s2b = (s) => s.length > 10;

const h: n2s = pipe(g, exclaim, isHuge);
// const j: n2s = compose(exclaim, f, g);

h(20);
```

Error

```
Type '(a: number) => boolean' is not assignable to type 'n2s'.
  Type 'boolean' is not assignable to type 'string'.
```

It seems TypeScript is confused about what types to expect - because we don't know what types to expect until we pass all our functions to the pipe, and it may be the case that none of the function types we pass end up being the correct type for the whole composition.

Reply:
Oops, no, this doesn't work, either. The type expectations are still too rigid. It fails even with the fallback:

```TypeScript
// eslint-disable no-explicit-any
// type a2a = (x: any) => any;
// const pipe = (...fns: a2a[]) => (x: any) => fns.reduce((y, f) => f(y), x);
type a2a = (x: any) => any;

function pipe<T1, R>(f1: (a: T1) => R): (a: T1) => R;
function pipe<T1, T2, R>(f1: (a: T1) => T2, f2: (a: T2) => R): (a: T1) => R;
function pipe<T1, T2, T3, R>(f1: (a: T1) => T2, f2: (a: T2) => T3, f3: (a: T3) => R): (a: T1) => R;
function pipe(...fns: a2a[]): a2a {
  return (input: any) => fns.reduce((prev, fn) => fn(prev), input);
}
// eslint-enable no-explicit-any

type n2n = (n: number) => number;
const g: n2n = n => n + 1;
const f: n2n = n => n * 2;
type n2s = (n: number) => string;
const exclaim:n2s = (n) => `${n}!`;
type s2b = (s: string) => boolean;
const isHuge:s2b = (s) => s.length > 10;
```

Error:
```
Errors in code
Argument of type 'n2s' is not assignable to parameter of type '(a: number) => number'.
  Type 'string' is not assignable to type 'number'.
```

Reply:
Note: [Ramda's type is 54 LOC of overloads](https://github.com/ramda/types/blob/develop/types/pipe.d.ts) and still breaks with this example. These problems have [plagued](https://github.com/typed-typings/npm-ramda/issues/29) all of the functional programming libraries for years when people tried to apply TypeScript to them.